test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ClientYamlTestExecutionContext.java

```diff
@@ -106,6 +106,46 @@ public ClientYamlTestResponse callApi(String apiName, Map<String, String> params
             requestParams.put(INCLUDE_TYPE_NAME_PARAMETER, "true");
         }
 
+        // When running tests against a mixed 7.x/6.x cluster we need to add the type to the document API
+        // requests if its not already included.
+        if ((apiName.equals("index") || apiName.equals("update") || apiName.equals("delete") || apiName.equals("get"))
+                && esVersion().before(Version.V_7_0_0) && requestParams.containsKey("type") == false) {
+            requestParams.put("type", "_doc");
+        }
+
+        // When running tests against a mixed 7.x/6.x cluster we need to add the type to the bulk API requests
+        // if its not already included. The type can either be on the request parameters or in the action metadata
+        // in the body of the request so we need to be sensitive to both scenarios
+        if (apiName.equals("bulk") && esVersion().before(Version.V_7_0_0) && requestParams.containsKey("type") == false) {
```

Comment:
Sorry if I'm missing something, but instead of this list of checks, could we just take the simpler approach of always doing `requestParams.put("type", "_doc");` when the request param is not already there? In particular, I'm not sure why we would only put the `type` parameter if `index` is also specified.

Reply:
We have to do this because in our tests we use both methods of specifying the index (and type) in a bulk request. Sometimes we set the index (and type) on the request params so its applied as a default and then the action metadata just contains `{index:{}}` or `{index:{ _id: foo}` and sometimes we don't specify the index (or type) on the request params and explicitly define them in the action metadata on each document we want to index. The problem with always setting the type on the request params is that if the index is not set on the request params the request will fail since its not valid to specify the type but not the index here.

Reply:
+1 to the current approach of this PR

Reply:
The way our request parsing works, I think you actually can specify the type without the index. In particular `POST _bulk?type=_doc { ... }` is a valid call.

It sounds like you and @jpountz are happy with this though, and this approach is sensible + straightforward, so I'm also happy if we want to merge.

Reply:
I think the issue here though is that the type and index are actually used in the path rather than queryParams on the URL (it seems in this class the two are put in the same map). So specifying a type without an index would try the endpoint `{type}/_bulk` which is not valid. I actually did it the way you suggested initially and found I got errors when I specified the type without the test specifying the index. 

As you are happy for me to merge I'll merge this now and have a closer look into this as a follow up